// graphql-lodash/lib/transformations.js
import every from 'lodash/every.js';
import some from 'lodash/some.js';
import startsWith from 'lodash/startsWith.js';
import endsWith from 'lodash/endsWith.js';
import lt from 'lodash/lt.js';
import lte from 'lodash/lte.js';
import gt from 'lodash/gt.js';
import gte from 'lodash/gte.js';
import eq from 'lodash/eq.js';
import map from 'lodash/map.js';
import keyBy from 'lodash/keyBy.js';
import chunk from 'lodash/chunk.js';
import drop from 'lodash/drop.js';
import dropRight from 'lodash/dropRight.js';
import take from 'lodash/take.js';
import takeRight from 'lodash/takeRight.js';
import flattenDepth from 'lodash/flattenDepth.js';
import fromPairs from 'lodash/fromPairs.js';
import nth from 'lodash/nth.js';
import reverse from 'lodash/reverse.js';
import uniq from 'lodash/uniq.js';
import uniqBy from 'lodash/uniqBy.js';
import countBy from 'lodash/countBy.js';
import filter from 'lodash/filter.js';
import reject from 'lodash/reject.js';
import groupBy from 'lodash/groupBy.js';
import sortBy from 'lodash/sortBy.js';
import minBy from 'lodash/minBy.js';
import maxBy from 'lodash/maxBy.js';
import meanBy from 'lodash/meanBy.js';
import sumBy from 'lodash/sumBy.js';
import join from 'lodash/join.js';
import get from 'lodash/get.js';
import mapValues from 'lodash/mapValues.js';
import at from 'lodash/at.js';
import toPairs from 'lodash/toPairs.js';
import invert from 'lodash/invert.js';
import invertBy from 'lodash/invertBy.js';
import keys from 'lodash/keys.js';
import values from 'lodash/values.js';
var transformations = {
    Array: {
        each: function (array, arg) {
            return map(array, function (item) { return applyTransformations(item, arg); });
        },
        map: map,
        keyBy: keyBy,
        chunk: chunk,
        drop: drop,
        dropRight: dropRight,
        take: take,
        takeRight: takeRight,
        flattenDepth: flattenDepth,
        fromPairs: fromPairs,
        nth: nth,
        reverse: reverse,
        uniq: uniq,
        uniqBy: uniqBy,
        countBy: countBy,
        filter: filter,
        reject: reject,
        filterIf: function (array, arg) {
            return filter(array, function (item) { return applyTransformations(item, arg); });
        },
        rejectIf: function (array, arg) {
            return reject(array, function (item) { return applyTransformations(item, arg); });
        },
        groupBy: groupBy,
        sortBy: sortBy,
        minBy: minBy,
        maxBy: maxBy,
        meanBy: meanBy,
        sumBy: sumBy,
        join: join,
    },
    Object: {
        get: get,
        mapValues: mapValues,
        at: at,
        toPairs: toPairs,
        invert: invert,
        invertBy: invertBy,
        keys: keys,
        values: values,
    },
    Number: {
        lt: lt,
        lte: lte,
        gt: gt,
        gte: gte,
        eq: eq,
    },
    String: {
        startsWith: startsWith,
        endsWith: endsWith,
    },
};
var opToExpectedType = {};
for (var type in transformations)
    for (var name_1 in transformations[type])
        opToExpectedType[name_1] = type;
export function applyTransformations(object, args) {
    if (!args)
        return object;
    for (var op in args) {
        if (object === null)
            break;
        var arg = args[op];
        if (op === 'and') {
            object = every(arg, function (predicateArgs) { return applyTransformations(object, predicateArgs); });
            continue;
        }
        if (op === 'or') {
            object = some(arg, function (predicateArgs) { return applyTransformations(object, predicateArgs); });
            continue;
        }
        var expectedType = opToExpectedType[op];
        var type = object.constructor && object.constructor.name;
        // handle objects created with Object.create(null)
        if (!type && (typeof object === 'object'))
            type = 'Object';
        if (expectedType !== type)
            throw Error("\"" + op + "\" transformation expect \"" + expectedType + "\" but got \"" + type + "\"");
        object = transformations[type][op](object, arg);
    }
    return object;
}
